File: app/frontend/src/types/thumbtack.ts
export type ThumbtackBusiness = 'holy_hauling' | 'holy_handy'

export interface ThumbtackConnection {
  id: string
  label: string
  city_id: string
  business: ThumbtackBusiness
  business_id: string | null
  url_token: string
  auth_username: string | null
  is_active: boolean
  last_event_at: string | null
  last_error_at: string | null
  created_at: string
}

/** Only ever returned by the create call — the secret is never fetchable again. */
export interface ThumbtackConnectionCreated extends ThumbtackConnection {
  webhook_url: string
  auth_secret: string
}

export interface ThumbtackEvent {
  id: string
  connection_id: string
  kind: 'lead' | 'message' | 'review' | 'unknown'
  external_id: string | null
  raw_body: string
  status: string
  error: string | null
  lead_id: string | null
  received_at: string
  processed_at: string | null
}

export interface ThumbtackConnectionCreate {
  label: string
  city_id: string
  business: ThumbtackBusiness
}

File: app/backend/app/schemas/thumbtack.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

Business = Literal["holy_hauling", "holy_handy"]


class ConnectionCreate(BaseModel):
    label: str = Field(min_length=1, max_length=120)
    city_id: str
    business: Business


class ConnectionPatch(BaseModel):
    is_active: Optional[bool] = None


class ConnectionOut(BaseModel):
    """Safe to return anywhere. Carries no secret and no hash."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    label: str
    city_id: str
    business: str
    business_id: Optional[str] = None
    url_token: str
    auth_username: Optional[str] = None
    is_active: bool
    last_event_at: Optional[datetime] = None
    last_error_at: Optional[datetime] = None
    created_at: datetime


class ConnectionCreated(ConnectionOut):
    """Returned exactly once, from the create call. The only time the plaintext
    secret exists outside Thumbtack's configuration."""

    webhook_url: str
    auth_secret: str


class EventOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    connection_id: str
    kind: str
    external_id: Optional[str] = None
    raw_body: str
    status: str
    error: Optional[str] = None
    lead_id: Optional[str] = None
    received_at: datetime
    processed_at: Optional[datetime] = None
